Full grade sync

darrensiegel opened the pull request:
Adds a "Grades Management" UI that instructors can use to update line items, sync grades, and export gradebook.

The UI is LiveView driven and allows the serial execution of a collection of tasks related to creating, updating line items and posting scores.  After completion of each "task" a progress bar in the UI is updated.

A key change that was also made was to introduce a prefix to the resourceId attribute that we are populating in the LMS created line item.  This helps to guarantee that another, non-Torus tool populating line items and setting resource_ids does not conflict with Torus. 

This feature uses LTI Names and Roles Provisioning service to determine all currently enrolled students so that we do not attempt to post scores for students that have been removed from the LMS section. 

A key limitation of the grade sync feature is that there is no way for it to know if a grade has been overridden in the LMS by an instructor (at least based on Canvas and the LTI_AGS testing I did).  Therefore, I changed this feature page specific and to include a warning regarding loss of custom grade adjustments. 

To access this new "Grades Management" UI, launch in from the LMS as an instructor.  On the Overview page at the top-right you will see a "Grades Management" link.

bitbldr commented:
Looks really great, I had a few things I wanted to bring up but may not necessarily be issues.

1. Trying to submit an assessment as an administrator/instructor results in a Not Authorized error which could be confusing
<img width="1031" alt="Screen Shot 2020-11-20 at 3 00 45 PM" src="https://user-images.githubusercontent.com/6248894/99918694-b2468b80-2ce6-11eb-9eca-6e320d98ba2c.png">

2. When I took an assessment as a student and got a "0%" or all incorrect, the score still shows up as a dash "-" as if I never took the assessment. I'm not sure if this is a canvas thing or a bug on our side
<img width="789" alt="Screen Shot 2020-11-22 at 4 48 57 PM" src="https://user-images.githubusercontent.com/6248894/99918730-dc984900-2ce6-11eb-854c-465070781f3c.png">
 
<img width="1072" alt="Screen Shot 2020-11-22 at 4 53 48 PM" src="https://user-images.githubusercontent.com/6248894/99918736-e621b100-2ce6-11eb-8ae5-059cdb04e4fa.png">

3. I accidentally left/reloaded the page after initiating a Grade Sync (then subsequently noticing the "do not leave this page" message). It left the gradebook in a strange state. Is it possible to store all the LTI details we need for grade passback an kick this process off in process outside of the live view so that if someone leaves the page this doesn't happen?
<img width="1102" alt="Screen Shot 2020-11-22 at 5 16 45 PM" src="https://user-images.githubusercontent.com/6248894/99918778-2a14b600-2ce7-11eb-99c7-557ebac199ff.png">

bitbldr commented on lib/oli/lti/lti_nrps.ex:
```diff
@@ -0,0 +1,74 @@
+defmodule Oli.Lti.LTI_NRPS do
```
Should this service live in the existing Oli.Lti_1p3 context? i.e. Oli.Lti_1p3.Nrps

darrensiegel replied:
My take on this was the `Oli.Lti_1p3` is a context for LTI v1.3 launching.  The NRPS and AGS aren't 1.3 specific.

darrensiegel commented:
Addressed the review comments:
1) I removed the restriction that was preventing instructors from submitting graded pages
2) I tracked down the bug that was causing 0% assessments from not showing up in the gradebook. This was a bug in the "best" scoring strategy implementation wrt how it calculates the "out_of" value.  I added a unit test assertion that exposes the bug and made the fix that allows the updated test to succeed.

darrensiegel commented:
Also adjusted the `submission_test.exs` since the change in the scoring strategy resulted in a different score being selected here as "the best".  Since in this test both parts were 100%, i adjusted it to avoid that state making the part 2 answer not 100%.
